src/System.Net.Http/src/System/Net/Http/HttpContent.cs

```diff
@@ -684,30 +674,41 @@ private static bool TryDetectEncoding(byte[] data, int dataLength, out Encoding
             return false;
         }
 
-        private static bool ByteArrayHasPrefix(byte[] byteArray, int dataLength, byte[] prefix)
+        private static bool BufferHasPrefix(ArraySegment<byte> buffer, byte[] prefix)
         {
-            if (prefix == null || byteArray == null || prefix.Length > dataLength || prefix.Length == 0)
+            byte[] byteArray = buffer.Array;
+            if (prefix == null || byteArray == null || prefix.Length > buffer.Count || prefix.Length == 0)
                 return false;
-            for (int i = 0; i < prefix.Length; i++)
+
+            for (int i = 0, j = buffer.Offset; i < prefix.Length; i++, j++)
             {
-                if (prefix[i] != byteArray[i])
+                if (prefix[i] != byteArray[j])
                     return false;
             }
+
             return true;
         }
 
         #endregion Helpers
 
-        private class LimitMemoryStream : MemoryStream
+        internal sealed class LimitMemoryStream : MemoryStream
         {
-            private int _maxSize;
+            private readonly int _maxSize;
+
+            public LimitMemoryStream(int maxSize)
```

Comment:
> This is just for testing, right?

No.

> Since 0 is a legal value for capacity to MemoryStream..ctor(int), why not just explicitly call the other ctor with 0 capacity?

Because that base ctor on MemoryStream does ```new byte[capacity]```, causing an array to be allocated unnecessarily. I contemplated adding a special-case to MemoryStream, but this seems like rare usage and that it would be better to handle it here rather than add a conditional there.